transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionMetadata.java

```diff
@@ -0,0 +1,663 @@
+                    boolean validTimestamp = txnStartTimestamp == transitMetadata.txnStartTimestamp() || allowedEmptyAbort;
+
+                    if (!validProducerEpoch(transitMetadata) ||
+                        !topicPartitions.equals(transitMetadata.topicPartitions()) ||
+                        txnTimeoutMs != transitMetadata.txnTimeoutMs() ||
+                        !validTimestamp) {
+                        throwStateTransitionFailure(transitMetadata);
+                    }
+                    break;
+
+                case COMPLETE_ABORT: // from write markers
+                case COMPLETE_COMMIT:
+                    if (!validProducerEpoch(transitMetadata) ||
+                        txnTimeoutMs != transitMetadata.txnTimeoutMs() ||
+                        transitMetadata.txnStartTimestamp() == -1) {
+                        throwStateTransitionFailure(transitMetadata);
+                    }
+                    break;
+
+                case PREPARE_EPOCH_FENCE:
+                    // We should never get here, since once we prepare to fence the epoch, we immediately set the pending state
+                    // to PrepareAbort, and then consequently to CompleteAbort after the markers are written.. So we should never
+                    // ever try to complete a transition to PrepareEpochFence, as it is not a valid previous state for any other state, and hence
+                    // can never be transitioned out of.
+                    throwStateTransitionFailure(transitMetadata);
+                    break;
+
+                case DEAD:
+                    // The transactionalId was being expired. The completion of the operation should result in removal of the
+                    // the metadata from the cache, so we should never realistically transition to the dead state.
+                    throw new IllegalStateException("TransactionalId " + transactionalId + " is trying to complete a transition to " +
+                        toState + ". This means that the transactionalId was being expired, and the only acceptable completion of " +
+                        "this operation is to remove the transaction metadata from the cache, not to persist the " + toState + " in the log.");
+
+                default:
+                    break;
+            }
+
+            LOGGER.debug("TransactionalId {} complete transition from {} to {}", transactionalId, state, transitMetadata);
+            producerId = transitMetadata.producerId();
+            prevProducerId = transitMetadata.prevProducerId();
+            nextProducerId = transitMetadata.nextProducerId();
+            producerEpoch = transitMetadata.producerEpoch();
+            lastProducerEpoch = transitMetadata.lastProducerEpoch();
+            txnTimeoutMs = transitMetadata.txnTimeoutMs();
+            topicPartitions = transitMetadata.topicPartitions();
```

Comment:
Line#135 uses `Set.copyOf` to create an immutable copy, which makes topicPartitions immutable. This could potentially lead a bug.

Perhaps we should use a `HashSet` instead to ensure mutability